### PR TITLE
make stepUp, stepDown, valueAsNumber, valueAsDate, and list work in HTMLInputElement

### LINKED ISSUE #19773: consider `min`, `max`, `step` for range type input element
need consider `min`, `max`, `step` for range type input element, when they are implemented
this will continue #19761

see 
- https://html.spec.whatwg.org/multipage/input.html#range-state-(type=range):value-sanitization-algorithm
-  https://html.spec.whatwg.org/multipage/#rules-for-parsing-floating-point-number-values

### LINKED ISSUE #25386: implement HTMLInputElement stepUp, stepDown,valueAsDate, valueAsNumber, list
https://html.spec.whatwg.org/multipage/input.html#htmlinputelement defines these and corresponding WPT tests under html/semantics/forms/the-input-element test them.

Servo also has commented-out IDL for valueHigh and valueLow; these were at one time proposed for use in `<input type="range" multiple>`, but the spec now says multiple doesn't apply to range and they can be removed entirely.

### LINKED ISSUE #25388: Setting HTMLInputElement's type to range (just to range, specifically) isn't updating its value
WPT html/semantics/forms/the-input-element/type-change-state.html tests various input element type-changes; Servo mostly passes, but fails when the new type is range.

The reason the test expects the new value to be 50 in particular is because, in the absence of other attributes, a range input goes from 0 to 100 and has a default value at the midpoint, as described in https://html.spec.whatwg.org/multipage/input.html#range-state-(type=range)

Rather than 50 or some other less-correct numeric value, we just continue to see the previously-set value, generally not a number at all.

### LINKED ISSUE #25390: Setting a range input's value isn't behaving itself
WPT html/semantics/forms/the-input-element/range-2.html tries putting various values into a range input element with the expectation that it will normalize them; the element is correctly turning 2.1 into 2, but it's getting various other cases wrong. This probably overlaps #25388 but doesn't seem exactly the same as it.

### PR DESCRIPTION
**Pull request by pshaughn**
Steppy operation of input elements now functions. A few related tests still fail because of incorrect test expectations. A couple range reftests fail on what might be correct expectations; solving that involves layout, not DOM.
Main changes here:
- str.rs now has its datetime format parsers public, with some changes to align their output with chrono::NaiveDateTime and fix a bug in milliseconds validation
- HTMLInputElement has new attributes, and code behind them for relevant concepts
- servoparser/mod.rs now has some extra code to ensure that input sanitization happens at the right moment during input element creation; the spec is non-specific about when this happens, but it has to happen somewhere and in #21952 @Eijebong had independently come to the same conclusion as me about the timing. 
- Some tests are added where WPT wasn't already testing functionality I was adding; Firefox and Chrome pass them, Safari fails some in a reasonably expected way
- Lots of test metadata changed to passing

One test continues to fail because 0.1 * 6 != 0.6 and the spec is written in abstractions instead of actual floating-point numbers; other test failures are WPT bugs I've reported as such and will write fixes for soon if no one else does.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #25390 and fix #25388 and fix #19773 (except validation doesn't exist) and fix #25386

- [X] There are tests for these changes OR

### CODE REVIEW

**Comment by pshaughn** [comment]
is that the wrong syntax?
@bors-servo try=wpt

**Comment by pshaughn** [comment]
List is working now, and I added a test for it... existing valueAsNumber tests aren't covering the chrono types, I'm gonna go in and add those too, and this is still a draft
@bors-servo try=wpt

**Comment by pshaughn** [comment]
Don't bother trying that push, it needs a rebase (because a WPT test was fixed upstream) and a bunch of idlharness test metadata.

**Comment by pshaughn** [comment]
I think this might be all synced up now.
@bors-servo try=wpt

File: tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/input-valueasnumber-stepping.html [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    <p><input type='datetime-local' id='input_datetime-local'></p>
+    <p><input type='range' id='input_range'></p>
+    <p><input type='number' id='input_number'></p>
+  </form>
+
+  <script>
+    function test_stepping(inputType, stringValue, steppedString, baseNumber, stepAmount) {
+       test(function() {
+          // put number in
+          input=document.getElementById("input_"+inputType);
```
Comment (jdm) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/input-valueasnumber-stepping.html]: Same nit comments as previous test files.

**Review by jdm — CHANGES_REQUESTED** [review]
This is great! There's a surprising amount of subtlety in the implementation here, and your comments make me feel more comfortable with revisiting this code in the future if the spec changes or incompatibilities warrant it.

File: components/script/dom/htmlinputelement.rs [hunk 2]
```diff
@@ -356,6 +376,330 @@ impl HTMLInputElement {
...
+
+                // Step 11
+                return self.SetValueAsNumber(value);
+            },
+        };
+    }
+
+    // https://html.spec.whatwg.org/multipage/#concept-input-list
+    fn suggestions_source_element(&self) -> Option<DomRoot<HTMLElement>> {
+        let listString = self
```
[the remark this answers is not in the document]
Comment (pshaughn) [reply components/script/dom/htmlinputelement.rs]: I'm not sure how that even happened. Is there a way test-tidy could be set up to catch it?

[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/htmlinputelement.rs]: I'm honestly surprised that the rust compiler warnings for this didn't fire.

[the remark this answers is not in the document]
Comment (jdm) [reply components/script/dom/htmlinputelement.rs]: Huh: https://github.com/servo/servo/blob/5c7a4db5f4e374dee287d403d90c7bc0e07ac9d0/components/script/lib.rs#L12

File: components/script/dom/htmlinputelement.rs [hunk 3]
```diff
@@ -356,6 +376,330 @@ impl HTMLInputElement {
...
+            .upcast::<Element>()
+            .get_string_attribute(&local_name!("list"));
+        if listString.len() > 0 {
+            let ancestors = self
+                .upcast::<Node>()
+                .inclusive_ancestors(ShadowIncluding::No);
+            if let Some(ancestor) = ancestors.last() {
```
[the remark this answers is not in the document]
Comment (pshaughn) [reply components/script/dom/htmlinputelement.rs]: Yes, it is and I hadn't realized that method existed at the time. I think I was copying from what I saw in CompareDocumentPosition, but that method had a reason to also want the entire list.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
```
  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 0000-12-10, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2019-00-12, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2019-12-00, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2019-13-10, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2019-02-29, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2019-12-10, expected valueAsNumber: 1575936000000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type date (actual value: 2016-02-29, expected valueAsNumber: 1456704000000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type date (actual valueAsNumber: 0, expected value: 1970-01-01)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type date (actual valueAsNumber: 1575936000000, expected value: 2019-12-10)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type month (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type month (actual value: 0000-12, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type month (actual value: 2019-00, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type month (actual value: 2019-12, expected valueAsNumber: 599)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type week (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type week (actual value: 0000-W50, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type week (actual value: 2019-W00, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type week (actual value: 2019-W60, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type week (actual value: 2019-W50, expected valueAsNumber: 1575849600000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type week (actual valueAsNumber: 0, expected value: 1970-W01)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: 24:00, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: 00:60, expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: 00:00, expected valueAsNumber: 0)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: 12:00, expected valueAsNumber: 43200000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type time (actual value: 23:59, expected valueAsNumber: 86340000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type datetime-local (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type datetime-local (actual value: 2019-12-10T00:00, expected valueAsNumber: 1575936000000)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type datetime-local (actual valueAsNumber: 1575936000000, expected value: 2019-12-10T00:00)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type number (actual value: , expected valueAsNumber: NaN)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type number (actual value: 123, expected valueAsNumber: 123)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type number (actual value: 123.456, expected valueAsNumber: 123.456)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type number (actual valueAsNumber: 123, expected value: 123)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: , expected valueAsNumber: 50)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: 0, expected valueAsNumber: 0)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: 50, expected valueAsNumber: 50)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: 100, expected valueAsNumber: 100)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: -10, expected valueAsNumber: 0)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber getter on type range (actual value: 110, expected valueAsNumber: 100)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type range (actual valueAsNumber: 0, expected value: 0)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type range (actual valueAsNumber: 50, expected value: 50)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”” PASS [expected FAIL] valueAsNumber setter on type range (actual valueAsNumber: 100, expected value: 100)

  â–¶ Unexpected subtest result in /html/semantics/forms/the-input-element/input-valueasnumber.html:
  â”‚ FAIL [expected PASS] valueAsNumber setter on type time (actual valueAsNumber: 86340000, expected value: 23:59)
  â”‚   â†’ assert_equals: expected "23:59" but got "23:59:00.000"
  â”‚ 
  â”‚ testValueAsNumberSetter/<@http://web-platform.test:8000/html/semantics/forms/the-input-element/input-valueasnumber.html:46:12
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1984:25
  â”‚ test@http://web-platform.test:8000/resources/testharness.js:548:30
  â”‚ testValueAsNumberSetter@http://web-platform.test:8000/html/semantics/forms/the-input-element/input-valueasnumber.html:43:8
  â”” @http://web-platform.test:8000/html/semantics/forms/the-input-element/input-valueasnumber.html:102:4
```

**Comment by pshaughn** [comment]
Interesting! It looks like that fail at the end is expecting the seconds to be left off the string value when they're zero, which I'm 99% sure is not specified anywhere and which has branches for it in other tests.

**Comment by pshaughn** [comment]
Sorry about that; the new tests from upstream led me to fix an actual bug despite the time-zone and optional-seconds test inconsistencies (one of my codepaths was multiplying seconds by 3600 instead of hours)
@bors-servo try=wpt

**Pull request merged**